### ISSUE TITLE
QtQuick fails in windowed onedir mode on macOS with Qt >= 5.12

### ISSUE THREAD

**Issue by cbenhagen** [pyinstaller/pyinstaller#4237]
Starting from PySide2/PyQt5 using Qt version 5.12.0 and higher, windowed onedir mode results in the following error:
```
QQmlApplicationEngine failed to load component
:5 Type ApplicationWindow unavailable
file:///private/var/folders/n0/1b5b642541s_q7rc6v0jhp_80000gp/T/pytest-of-ben/pytest-4/test_PySide2_QtQuick_onedir_0/dist/test_PySide2_QtQuick.app/Contents/Resources/PySide2/qml/QtQuick/Controls.2/qmldir:-1 module ".private.var.folders.n0.1b5b642541s_q7rc6v0jhp_80000gp.T.pytest-of-ben.pytest-4.test_PySide2_QtQuick_onedir_0.dist.test_PySide2_QtQuick.app.Contents.Resources.PySide2.qml.QtQuick.Controls" plugin "qtquickcontrols2plugin" not found
```

Older versions of PySide2 and PyQt5 work. @bjones I see you have recently worked on the PySide2 and PyQt5 hooks. Do you think we should only test windowed mode? Or does it make sense to test Qt apps in console mode?

Any ideas to make the tests nicer than with this copy paste from the QWebEngine test?

```diff
diff --git a/tests/functional/test_libraries.py b/tests/functional/test_libraries.py
index 78fd00f5..9520f1a1 100644
--- a/tests/functional/test_libraries.py
+++ b/tests/functional/test_libraries.py
@@ -307,8 +307,7 @@ def test_PyQt5_QWebEngine(pyi_builder, data_dir):
 @PYQT5_NEED_OPENGL
 @importorskip('PyQt5')
 def test_PyQt5_QtQml(pyi_builder):
-    pyi_builder.test_source(
-        """
+    source_to_test = """
         import sys
 
         from PyQt5.QtGui import QGuiApplication
@@ -340,7 +339,28 @@ def test_PyQt5_QtQml(pyi_builder):
         res = app.exec_()
         del engine
         sys.exit(res)
-        """)
+        """
+
+    if is_darwin:
+        # 1. Only test the Mac .app bundle, by modifying the executes this
+        #    fixture runs.
+        _old_find_executables = pyi_builder._find_executables
+        # Create a replacement method that selects just the .app bundle.
+
+        def _replacement_find_executables(self, name):
+            path_to_onedir, path_to_app_bundle = _old_find_executables(name)
+            return [path_to_app_bundle]
+        # Use this in the fixture. See https://stackoverflow.com/a/28060251 and
+        # https://docs.python.org/3/howto/descriptor.html.
+        pyi_builder._find_executables = \
+            _replacement_find_executables.__get__(pyi_builder)
+
+        # 2. Run the test with specific command-line arguments.
+        pyi_builder.test_source(source_to_test,
+                                pyi_args=['--windowed'])
+    else:
+        # The Linux and Windows test needs no special handling.
+        pyi_builder.test_source(source_to_test)
 
 
 @importorskip('PyQt5')
@@ -389,8 +409,7 @@ def test_PySide2_QWebEngine(pyi_builder):
 
 @importorskip('PySide2')
 def test_PySide2_QtQuick(pyi_builder):
-    pyi_builder.test_source(
-        """
+    source_to_test = """
         import sys
 
         # Not used. Only here to trigger the hook
@@ -419,7 +438,28 @@ def test_PySide2_QtQuick(pyi_builder):
         QTimer.singleShot(0, app.exit)
 
         sys.exit(app.exec_())
-        """)
+        """
+
+    if is_darwin:
+        # 1. Only test the Mac .app bundle, by modifying the executes this
+        #    fixture runs.
+        _old_find_executables = pyi_builder._find_executables
+        # Create a replacement method that selects just the .app bundle.
+
+        def _replacement_find_executables(self, name):
+            path_to_onedir, path_to_app_bundle = _old_find_executables(name)
+            return [path_to_app_bundle]
+        # Use this in the fixture. See https://stackoverflow.com/a/28060251 and
+        # https://docs.python.org/3/howto/descriptor.html.
+        pyi_builder._find_executables = \
+            _replacement_find_executables.__get__(pyi_builder)
+
+        # 2. Run the test with specific command-line arguments.
+        pyi_builder.test_source(source_to_test,
+                                pyi_args=['--windowed'])
+    else:
+        # The Linux and Windows test needs no special handling.
+        pyi_builder.test_source(source_to_test)
 
 
 @importorskip('zope.interface')
```

```
PyInstaller: 3.5.dev0+gd7405248
Python: 3.7.3
Platform: Darwin-18.6.0-x86_64-i386-64bit (macOS Mojave 10.14.5)
```

**Comment by bjones1** [comment]
@cbenhagen, thanks for taking a look at this. I'd like to fix it, but I don't have access to a Mac. I'll take a look at your changes in a week or so -- I'm on vacation and have very limited Internet access. More as soon as I can!

**Comment by bjones1** [comment]
So, to follow up a bit, this only failed in windowed onedir? That is, console onedir runs just fine?

**Comment by cbenhagen** [comment]
Yes onedir works like expected.

**Comment by bjones1** [comment]
OK, I was able to reproduce this and fix it. I'll work on adding test cases / avoiding a lot of QWebEngine test case copies.

**Comment by bjones1** [comment]
This should now be fixed in 4e284ac346fe98605eeeae577ac97f3a08b9f46d -- would you check?

**Comment by cbenhagen** [comment]
Thanks!
